syntax = "proto3";

package raft;

option go_package = "/raft";

message RequestVoteRequest {
  int32 term = 1;
  int32 candidate_id = 2;
}

message RequestVoteResponse {
  int32 term = 1;
  bool vote_granted = 2;
}

message LogEntry {
  string operation = 1;
  int32 term = 2;
  int32 index = 3;
}

message AppendEntriesRequest {
  int32 term = 1;
  int32 leader_id = 2;
  repeated LogEntry entries = 3;
  int32 commit_index = 4;
}

message AppendEntriesResponse {
  int32 term = 1;
  bool success = 2;
}

service Raft {
  rpc RequestVote(RequestVoteRequest) returns (RequestVoteResponse);
  rpc AppendEntries(AppendEntriesRequest) returns (AppendEntriesResponse);
}
